{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e91f1588-81ed-419c-9ca4-96e76f1979ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install PyMuPDF\n",
    "!pip install spacy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "df781bfa-9211-4ec6-9c3b-7eb63ea8ec83",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-12T10:20:54.955962Z",
     "start_time": "2025-10-12T10:20:53.387487Z"
    }
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "import fitz  # PyMuPDF for PDF parsing\n",
    "import spacy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dfddceb-43c2-4ccb-8054-6e5628e57c67",
   "metadata": {},
   "outputs": [],
   "source": [
    "!python -m spacy download en_core_web_sm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "967ca801-a675-4ba9-be7a-dbeb919ed7ad",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-12T10:21:05.746211Z",
     "start_time": "2025-10-12T10:21:05.741149Z"
    }
   },
   "outputs": [],
   "source": [
    "def parse_pdf(path):\n",
    "    fitzed_doc = fitz.open(path)\n",
    "    text = \"\"\n",
    "    for page in fitzed_doc:\n",
    "        text += page.get_text()\n",
    "    return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aea5f353-6f27-4baa-814e-8f6c0ed2f942",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-12T10:21:58.815028Z",
     "start_time": "2025-10-12T10:21:58.583073Z"
    }
   },
   "outputs": [],
   "source": [
    "nlp = spacy.load(\"en_core_web_sm\")\n",
    "paper = r\"Papers/scitranslmed.adi0673 annotated.pdf\"  # Path to the PDF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dd356560-b3b2-4e94-8237-6820a8f03cb5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-12T10:22:03.171308Z",
     "start_time": "2025-10-12T10:22:00.094102Z"
    }
   },
   "outputs": [],
   "source": [
    "raw_text = parse_pdf(paper)\n",
    "doc = nlp(raw_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f4632c3-a8e6-44e7-976b-016def597ce8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-12T11:14:28.248980Z",
     "start_time": "2025-10-12T11:14:28.241904Z"
    }
   },
   "outputs": [],
   "source": [
    "hypothesis_patterns = (\n",
    "    r\"(hypothesis|H\\d+)[:–-]\",\n",
    "    r\"(theor(ies|y)|hypothes[ei]s) (is|was|are|were|that)\",\n",
    "    r\"(we|I|authors?|it is|study|theor(ies|y)|hypothes[ei]s) (hypothesi[sz]e|predict|propose|assume|expect|anticipate|foresee|postulate|conjecture|test|suggest)\",\n",
    "    r\"(hypothesi[sz]e|predict|propose|assume|expect|anticipate|foresee|postulate|aim|conjecture|test|suggest)(s|d)? that\"\n",
    "    r\"(investigate|examine|predict|test)(s|d)? (whether|if)\",\n",
    "    r\"(evidence|data) (predict|suggest|propose|assume)\"\n",
    "    r\"(aim|objective|goal) (of th(is|e) study)\",\n",
    "    r\"if.*then\",\n",
    "    r\"it would appear\",\n",
    "    r\"it likely that\",\n",
    "    r\"According to the (theor(ies|y)|hypothes[ei]s|views?)\"\n",
    ")\n",
    "\n",
    "methods_patterns = (\n",
    "    r\"(methods?|methodology|procedures?|techniques?) (is|was|are|were|that|[:–-])\",\n",
    "    r\"(we|I|authors?|the study) (used|employed|applied|conducted|performed|carried out|implemented)\",\n",
    "    r\"(data (collection|gathering|acquisition)|samples?|participants?|subjects?) (were|was)\",\n",
    "    r\"(measurements?|assays?|tests?|analyses?) (were|was) (performed|conducted|carried out|done)\",\n",
    "    r\"(statistical|analysis|analytical) (methods?|approach|procedure)\",\n",
    "    r\"(in order to (determine|measure|assess|evaluate|test|quantify))\",\n",
    "    r\"(equipment|instruments?|software|apparatus|tools?) (was|were|used)\",\n",
    "    r\"(protocol|procedure|experimental design) (was|were)\",\n",
    "    r\"(collection period|duration|timeframe) (was|of)\",\n",
    "    r\"(participants?|subjects?|samples?) (were |was )?(recruited|selected|enrolled|obtained)\",\n",
    "    r\"(inclusion|exclusion) (criteria|requirements)\",\n",
    "    r\"(we|I|authors?) (measure|perform|conduct|test|quantif(y|ied)|employ|adopt)\",\n",
    "    r\"(was|were) (stained|embedded|measured|performed|calculated|obtained|identified|detected|monitored|isolated|stimulated|used|removed|harvested|acqired|gated|tested|considered)\",\n",
    "    r\"(regression|ANOVA|t-test|chi-square|Wilcoxon|Mann-Whitney|Kruskal-Wallis|Fisher['’]s (exact)?) (test|analysis|model)\",\n",
    "    r\"(correlation|association|relationship) (was|were) (assessed|examined|evaluated|tested)\",\n",
    "    r\"(clustering|classification|regression|neural network|deep learning|random forest|support vector) (was|were) (performed|applied|used)\",\n",
    "    r\"(principal component analysis|PCA|factor analysis|dimensionality reduction) (was|were) (performed|conducted)\",\n",
    "    r\"(R|Python|MATLAB|SPSS|SAS|Stata|GraphPad|TensorFlow|PyTorch|scikit-learn) (was|were) used\",\n",
    "    r\"(software|tool|package|library) (was|were) used\",\n",
    "    r\"(cross-validation|bootstrapping|permutation test) (was|were) (performed|conducted|used)\",\n",
    ")\n",
    "\n",
    "dataset_patterns = (\n",
    "    r\"(dataset|data set|database|data source|repository) (was|were|is|are|used|obtained|available|collected|generated|produced)\",\n",
    "    r\"(available|accessible|open) (dataset|data set|database|data source|repository)\",\n",
    "    r\"(downloaded|retrieved|obtained|accessed) from (the )?(database|repository|server|(web)?site|portal)\",\n",
    "    r\"data( were)? (obtained|retrieved|downloaded|sourced)\",\n",
    "    r\"(doi|accession number|repository|github|zenodo|figshare|osf|dryad)\\s*[:–-]\",\n",
    "    r\"(supplementary )?data (available )?(at|in|from|on)\",\n",
    "    r\"(gene expression|genomic|proteomic|metabolomic|imaging) dataset\",\n",
    "    r\"(data (necessary )?to reproduce|reproducib(le|ility))\",\n",
    "    r\"(accession|catalog) (number|code)\\s*[:–-]\",\n",
    "    r\"(data|dataset) (was|were) (normalized|standardized|preprocessed|filtered|transformed|cleaned)\",\n",
    "    r\"(normalized|standardized|preprocessed|filtered|transformed|cleaned) (the )?(data|dataset)\",\n",
    ")\n",
    "\n",
    "experiment_patterns = (\n",
    "    r\"(experiment|assay|trial|test)\\s*(\\d+|[A-Z])?(\\s*[:–-])\",\n",
    "    r\"(test|examine|investigate|evaluate|assess|determine|measure) (the )?(hypothesis|prediction|effect|impact|role|function)\",\n",
    "    r\"(performed|conducted|carried out|designed|set up) (an? )?(experiment|assay|trial|test|study)\",\n",
    "    r\"experimental (design|setup|procedure|protocol|approach)\",\n",
    "    r\"(treatment|condition|group) (was|were)( exposed| treated)\",\n",
    "    r\"(control|experimental|treatment) group\",\n",
    "    r\"(exposed|subjected|treated|stimulated|challenged) (to|with)\",\n",
    "    r\"(intervention|treatment|manipulation) (was|were)\",\n",
    "    r\"(validation|confirmatory|confirmation) (experiment|assay|test)\",\n",
    "    r\"(parallel|sequential|longitudinal) (experiment|test|measurement)\",\n",
    ")\n",
    "\n",
    "analysis_patterns = (\n",
    "    r\"(analysis|analyses) (was|were|performed|conducted)\",\n",
    "    r\"(p-value(s)?|statistical significance) (was|were) (calculated|assessed|determined)\",\n",
    "    r\"(supervised|unsupervised) (learning|modeling|analysis) (was|were)\",\n",
    "    r\"(accuracy|precision|recall|F1 score|AUC|ROC curve) (was|were|are|is)\",\n",
    "    r\"(we|I|authors?) (evaluated|assessed|validated) (the )?(model|results|performance|hypothesis)\",\n",
    "    r\"(data|results) (was|were) (visualized|plotted|presented|graphed) (using|with|as)\",\n",
    ")\n",
    "\n",
    "result_patterns = (\n",
    "    r\"(we|I|authors?) (analyzed|assessed|evaluated|quantified|examined|compared) (the )?(data|results|effects)\",\n",
    "    r\"(results?|findings?|outcomes?|data|evidence|analysis) (show|demonstrate|indicate|reveal|suggest|support|imply|point|provide|confirm|validate)\",\n",
    "    r\"(we|I|authors?|the study) (found|observed|noted|reported|detected|identified|demonstrate|confirm|establish|conclude|infer)\",\n",
    "    r\"((no )?significant|(not )?statistically significant) (effect|difference|correlation|association|relationship) (was|were) (found|observed|detected)\",\n",
    "    r\"(p (<|>|<=|>=|=) \\d*\\.?\\d+| p-value(s)? (of|less than|greater than))\",\n",
    "    r\"(higher|lower|greater|smaller|increased|decreased|stronger|weaker) (than|compared to|relative to)\",\n",
    "    r\"(positive|negative|significant|weak|strong) (correlation|association|relationship) (was|were) (found|observed|detected)\",\n",
    "    r\"(trend(s)?|pattern(s)?) (was|were) (observed|noted|evident|apparent)\",\n",
    "    r\"(model|algorithm) (achieved|obtained|yielded|produced) (an? )?(accuracy|precision|recall|F1 score|AUC)\",\n",
    "    r\"(effect|impact|influence|role) (of \\w+ on \\w+) (was|were) (observed|found|significant|notable)\",\n",
    ")\n",
    "\n",
    "compiled_hypothesis_patterns = [re.compile(p, re.IGNORECASE) for p in hypothesis_patterns]\n",
    "compiled_methods_patterns = [re.compile(p, re.IGNORECASE) for p in methods_patterns]\n",
    "compiled_dataset_patterns = [re.compile(p, re.IGNORECASE) for p in dataset_patterns]\n",
    "compiled_experiment_patterns = [re.compile(p, re.IGNORECASE) for p in experiment_patterns]\n",
    "compiled_analysis_patterns = [re.compile(p, re.IGNORECASE) for p in analysis_patterns]\n",
    "compiled_result_patterns = [re.compile(p, re.IGNORECASE) for p in result_patterns]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6665bb7b-b630-4a4f-a327-7e5b2dc2ab60",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-12T11:14:32.302108Z",
     "start_time": "2025-10-12T11:14:32.151246Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['We hypothesized that the existence of high-\\xadaffinity peanut-\\xadspecific IgG+ memory B cells and their abil- ity to undergo class switching to IgE are critical for allergy persistence in pediatric peanut allergy.',\n",
       "  'In line with our model of IgE cell differentiation developed from mouse studies (2, 3, 7, 43), we proposed that allergen-\\xadspecific IgG+ memory B cells ex- pressing CD23/FCER2 and germline IGHE are precursors of patho- genic IgE+ plasma cells in highly sensitized peanut-\\xadallergic individuals.',\n",
       "  'We propose that CD23+IGHE+IgG1+ memory B cells are involved in the persistence of food allergy by providing precursors of pathogenic IgE plasma cells.',\n",
       "  'We hypothesized that IgG1+ memory B cells are the precursors of high-\\xadaffinity, pathogenic IgE+ plasma cells in persistent peanut allergy.',\n",
       "  '10X Genomics scRNA-\\xadseq processing and analysis 10X Genomics 5′ scRNA-\\xadseq data were processed with Cell Ranger version 3.1 and aligned to refdata-\\xadcellranger-\\xadGRCh38-\\xad3.0.0.',\n",
       "  'We tested this in our data by quantifying the SHM frequency of IGHM heavy chain consensus sequences.',\n",
       "  'M. Descatoire, S. Weller, S. Irtan, S. Sarnacki, J. Feuillard, S. Storck, A. Guiochon-\\xadMantel, J. Bouligand, A. Morali, J. Cohen, E. Jacquemin, M. Iascone, C. Bole-\\xadFeysot, N. Cagnard, J. C. Weill, C. A. Reynaud, Identification of a human splenic marginal zone B cell precursor with NOTCH2-\\xaddependent differentiation properties.'],\n",
       " ['A corresponding population of “type 2–marked” IgG1+ memory B cells was identified in single-\\xadcell RNA sequencing experiments.',\n",
       "  'CD27 was used as a marker of memory B cells that are somatically mutated and likely to contain allergen-\\xadspecific clones (26).',\n",
       "  'A cluster of somatically mutated switched memory cells expresses FCER2, IL4R, and IGHE To further characterize B cell memory subsets, we performed 10X Genomics scRNA-\\xadseq of whole transcriptomes and BCRs on flow-\\xad sorted CD27+ memory B cells from five peanut-\\xadallergic children with PsIgE>100 and three non-\\xadallergic children (table S2).',\n",
       "  'To further understand the biological processes shaping the memory B cells in cluster 5, we performed a Gene Ontology (GO) enrich- ment analysis.',\n",
       "  'These conditions were op- timized for B cell proliferation and plasma cell differentiation and also allowed class switching to IgE. After 7 days, peanut-\\xadspecific anti- bodies were measured in the culture supernatants, and cells were analyzed by flow cytometry.',\n",
       "  'In day 9 culture supernatants, peanut-\\xadspecific IgG was detected in 93% of wells from CD23+IgG+ memory B cells but only in 12% of wells from CD23−IL-\\xad4R+IgG+ memory B cells (Fig. 4C).',\n",
       "  'DT binders were detected at similar frequencies among the three groups (fig. S10B).',\n",
       "  'i c l e 6 of 14 (cluster 5; Fig. 2), we used real-\\xadtime polymerase chain reaction (PCR) to quantify FCER2/CD23, germline IGHE (GLT-\\xadIGHE), and total-\\xad\\u200bIGHE transcripts in all antigen-\\xadbinding IgG1+ cells from peanut-\\xadallergic individuals with PsIgE>100 and from non-\\xadallergic individuals (Fig. 5C).',\n",
       "  'FCER2/CD23 expression was detected in most Ara h 2–binding IgG1+ cells from peanut-\\xadallergic individuals, and its average expression was significantly higher in Ara h 2–binding IgG1+ cells than DT-\\xadbinding IgG1+ cells from peanut-\\xadallergic (P = 0.0028) or non-\\xadallergic individuals (P = 0.0150).',\n",
       "  'Total-\\xad\\u200bIGHE and germline IGHE transcripts were detected in most Ara h 2–binding cells but few DT-\\xadbinding cells.',\n",
       "  'High-\\xadaffinity convergent Ara h 2–specific BCR sequences were identified across different individuals BCRs within individual humans are highly diverse, and identification of similar sequences among different individuals may be indicative of convergent antigen-\\xaddriven selection.',\n",
       "  'To identify convergent BCRs specific for Ara h 2, we performed clustering on the basis of IGH amino acid sequence similarity using the BCR sequencing data from Ara h 2 and DT binders and from all memory B cells profiled in the 10X scRNA-\\xadseq experiment.',\n",
       "  '(B) Culture supernatants of IgG+ memory B cells from non-\\xadallergic individuals (N = 7, left), peanut-\\xadallergic individuals with PsIgE<5 (N = 7, middle), and peanut-\\xadallergic individuals with PsIgE>100 (N = 7, right) were tested for the reac- tivity against peanut by ELISA.',\n",
       "  '(C) Culture supernatants of IgG+ memory B cells from peanut-\\xadallergic individuals with PsIgE>100 (N = 4) were tested for the reactivity of IgG against peanut (top) and Ara h 2 (bottom) by ELISA.',\n",
       "  'Al- though this was considered evidence of lower maturation of IgE+ cells, it could reflect a developmental relationship between IgE+ cells and their potential precursors, CD23+IgG1+ memory B cells.',\n",
       "  'The scRNA-\\xadseq experiment was performed with a relatively small number of samples and did not provide enough power for group comparisons.',\n",
       "  'The analysis was performed with PBMC samples from blood collected between 2019 and 2023.',\n",
       "  'Blood samples that yielded more than 30 × 106 PBMCs were used for the study.',\n",
       "  'Thus, we performed a phe- notypic and transcriptional analysis of total memory B cells by flow cytometry and 10X Genomics (WTA and BCR) scRNA-\\xadseq and characterized allergen-\\xadspecific B cells by plate-\\xadbased single-\\xadB cell analysis of BCR sequence and selected genes’ expression.',\n",
       "  'Each experimental value was obtained from a unique sample, because only one blood sample was obtained from each donor.',\n",
       "  'Patient samples For flow cytometry and the initial cell culture studies, PBMCs were isolated from peripheral blood from 58 pediatric patients with peanut allergy and 13 non-\\xadallergic children.',\n",
       "  'PsIgE concentrations obtained on the day of the blood draw were used to select the patients.',\n",
       "  'Four samples were obtained from PsIgE>100 pa- tients for the additional cell culture studies.',\n",
       "  '10X Genomics scRNA-\\xadseq of memory B cells B cells were isolated from 17 million to 100 million PBMCs from five peanut-\\xadallergic individuals with sIgE>100 and three non-\\xadallergic individuals (table S2) by EasySep Human B cell Isolation Kit (STEM- CELL Technologies, #17954).',\n",
       "  'V, D, and J genes from the IGH, IGL, and IGK loci were removed so that the properties of the BCR expressed by the cell would remain independent of the cluster it was assigned to.',\n",
       "  'Integration was performed using the previously identified top variable genes of each run and the first 20 dimensions.',\n",
       "  'i c l e 11 of 14 identified as T cells and naïve B cells, which were removed.',\n",
       "  'To identify DEGs in each cluster, the FindAllMarkers function from Seurat was used.',\n",
       "  'P values were calculated using a Wilcoxon rank sum test and were adjusted using a Bonferroni adjustment.',\n",
       "  'To identify the enrichment of GOs for each cluster, the list of positive DEGs for each cluster was tested using enrichR v3.0 (58, 59).',\n",
       "  'Similarly, detoxified DT (CRM197, PFENEX Biopharmaceuticals) was used to generate DT multimers.',\n",
       "  'Pre-\\xadenriched B cells by EasySep Human B cell Isolation Kit (STEMCELL Technologies, #17954) from eight peanut-\\xadallergic individuals with PsIgE>100, five peanut-\\xad allergic individuals with PsIgE<5, and two non-\\xadallergic individuals were used for Ara h 2 binder or DT binder sorting (table S5).',\n",
       "  'Plates were thawed on ice, and then oligo dT was added and annealing was performed at 70°C for 5 min. cDNA synthesis and cDNA amplification were performed with template switching oligo (TSO) using template switching RT enzyme mix (NEB, #M0466) according to the manufacturer’s instruc- tion.',\n",
       "  'The first PCR was performed to amplify BCR heavy and light chains using outer primers targeting human immunoglobulin con- stant regions (IgG and IgM for heavy chains; kappa and lambda for light chains) and TSO inner primer (table S6), respectively.',\n",
       "  'All PCR reactions were performed using the NEBNext Ultra II Q5 master mix (NEB, #M0544).',\n",
       "  'B cell receptor sequence processing and analysis BCR sequencing was obtained from two sources: 10X Genomics scRNA-\\xad seq + BCR sequencing and PacBio single-\\xadcell BCR sequencing.',\n",
       "  'Nonproductive heavy and light chain BCR sequences were removed.',\n",
       "  'Further processing was performed using presto v0.6.2 (62).',\n",
       "  'To remove reads too short to likely contain functional BCRs and associated constant regions, reads shorter than 500 bp were removed.',\n",
       "  'Nucleotide sites matching primer sequences, as well as the sites downstream of them, were removed.',\n",
       "  'Sites in which >50% of reads contained gap characters were removed.',\n",
       "  'Read groups containing fewer than 10 reads or in which >10% of sites differed from their final consensus sequence were removed.',\n",
       "  'Nonproductive heavy and light chain BCR sequences were removed.',\n",
       "  'As a further quality control, 16 sequences originated from wells in which no Ig DNA was detected through PCR and were discarded.',\n",
       "  'For ambiguous V gene assignments, only the first assignment was used for reconstruction.',\n",
       "  'The Shapiro-\\xadWilk test was performed to select either one-\\xadway ANOVA or the Kruskal-\\xadWallis test for the statistical significance for >2 groups.',\n",
       "  'Either the two-\\xadsided Student’s t test or the two-\\xadsided Mann-\\xadWhitney U test was used to perform the parametric or nonparametric tests, respectively, for two-\\xadgroup comparisons with a 95% confidential inter- val for the median.',\n",
       "  'For correlation analysis, the Shapiro-\\xadWilk test was performed to select a nonparametric Spearman’s correlation test.',\n",
       "  'The chi-\\xadsquare test was used to determine the significant differ- ences between expected frequencies and observed frequencies.'],\n",
       " ['Data in (D) are presented as box and whiskers plots in which boxes represent the median as well as the first and third quartiles, and whiskers extend to the largest and smallest values within 1.5 times the interquartile range from the boxes.',\n",
       "  'We then used Seurat’s IntegrateData function to combine data from the three sequencing runs.',\n",
       "  'BCR sequence data from 10X Genomics scRNA-\\xadseq + BCR sequencing began with the filtered V(D)J contigs from 10X Genomics Cell Ranger version 3.1.',\n",
       "  'For BCR sequencing data from sorted single-\\xadcell PacBio sequencing data, adaptor ligation, sequencing reactions, and initial sequence processing were conducted at GENEWIZ, LLC/Azenta US Inc. using standard PacBio circular consensus sequence tools and guidelines.',\n",
       "  'For BCR data from both data sources, SHM frequencies were de- termined for each cell as the frequency of nonambiguous mismatch- es along the immunoglobulin heavy chain variable (IGHV) gene of the unmutated germline sequence along IMGT positions 1 to 312 using shazam v1.1.2 (65).',\n",
       "  'S1 to S13 Tables S1 to S7 Other Supplementary Material for this manuscript includes the following: Data files S1 to S4 MDAR Reproducibility Checklist REFERENCES AND NOTES 1. S. H. Sicherer, H. A. Sampson, Food allergy: A review and update on epidemiology, pathogenesis, diagnosis, prevention, and management.'],\n",
       " ['L. conceptualized the study, and M.O. designed the experimental approach.'],\n",
       " ['Our transcriptional analysis was done on CD27+ memory B cells, and it did not distinguish CD45 isoforms.'],\n",
       " ['Our findings indicate that CD23+IgG1+ memory B cells transcribing germline IGHE are a unique memory population containing precur- sors of high-\\xadaffinity pathogenic IgE-\\xadproducing cells that are likely to be involved in the long-\\xadterm persistence of peanut allergy.',\n",
       "  'Consistent with previous observations, we found a significantly (P < 0.001) higher percentage of CD23+ total, naïve, and memory B cells in peanut-\\xad allergic than in non-\\xadallergic children of a similar age range (Fig.',\n",
       "  'A significant correlation was found between to- tal IgE and CD23+IgM/IgD+ memory B cells (P = 0.004), CD23+ IgM/IgD− memory B cells (P < 0.001), and CD23+IgG+ memory B cells (P < 0.001) (Fig. 1B).',\n",
       "  'The concentrations of total IgE in circula- tion also correlated significantly (P < 0.001) with PsIgE and with IgE specific for the main peanut allergen, Ara h 2 (fig.',\n",
       "  'A significant (P < 0.001) difference in total IgE concentrations was observed between individuals with a very high concentration of PsIgE in the blood (PsIgE>100 kUA/liter referred to as PsIgE>100, n = 18) and those with PsIgE lower than 5 kUA/liter (referred to as PsIgE<5, n = 16), whereas total IgE concentrations varied among peanut-\\xadallergic individuals with intermediate PsIgE (5 < PsIgE<100) (fig. S1B).',\n",
       "  'However, only the PsIgE>100 peanut-\\xad allergic group had a significantly higher frequency of CD23+ IgM/ IgD− memory B cells (P < 0.001) and CD23+ IgG+ memory B cells (P = 0.002) (Fig. 1D and fig. S2D) without a significant (P > 0.05) increase in total IgG+ memory B cells (fig.',\n",
       "  'Although the frequency of CD23+ naïve B cells was similarly increased in peanut-\\xadallergic individuals with very high or very low concentrations of PsIgE, only highly sensitized individuals had a significantly higher frequency of CD23+CD27+IgG+ memory B cells as percentage of total B cells or of CD27+ memory B cells compared with non-\\xadallergic individuals (P = 0.0002 in Fig.',\n",
       "  '1D and P = 0.004 in fig. S3C).',\n",
       "  'We identified 10 clusters of memory B cells (numbered 0 to 9) and identified dif- ferentially expressed genes (DEGs) in each cluster (Fig. 2, A and B, figs.',\n",
       "  'We detected CD73/ NT5E expression in only a small percentage of cells in clusters 4, 5, and 6, and transcripts for FAS/CD95 and ITGAX/CD11c were undetect- able in most cells (fig. S7).',\n",
       "  'We found that cells from peanut-\\xadallergic individuals in cluster 5 were enriched for cytokine-\\xadmediated signaling pathway signatures (GO: 009221) (ta- ble S4).',\n",
       "  'Although true IgE cells were rare, we identified two B cells with IGHE BCRs, each from a different peanut-\\xadallergic individual and both belonging to cluster 5 (fig. S8).',\n",
       "  'In the scRNA-\\xadseq experiment, we identified 178 B cells with normalized IGHE transcript values at least as high as the two B cells found with IGHE BCRs (Fig. 3A).',\n",
       "  'Genes are positively DEGs (adjusted P < 0.05) for cluster 5.',\n",
       "  'Enrichment for IGHG-\\xadassociated BCRs was ana- lyzed by chi-\\xadsquare test (cluster 5: 61.7% of cells, P < 2.2 × 10−16).',\n",
       "  'Class switching to IgE was relatively low in these culture conditions; nevertheless, we detected the highest frequency of IgE+ B cells in the cultures of CD23+/IL-\\xad4R+IgG+ memory B cells from peanut-\\xadallergic children with high PsIgE (PsIgE>100) (fig. S9C).',\n",
       "  'Total IgE was significantly (P < 0.001) higher in supernatants from CD23+IgG+ memory B cells than in supernatants from CD23−IL-\\xad4R+ (or CD23−IL-\\xad4R−)',\n",
       "  'The frequency of Ara h 2–binding B cells was significantly (P = 0.02) higher in peanut-\\xadallergic individuals with PsIgE>100 than in peanut-\\xadallergic individuals with PsIgE<5, whereas Ara h 2 binders were undetectable in non-\\xadallergic samples.',\n",
       "  'Heavy chain sequences of Ara h 2–binding IgG+ B cells from peanut-\\xad allergic individuals were significantly (P < 0.001) more mutated than DT-\\xadbinding IgG+ B cells from both peanut-\\xadallergic and non-\\xad allergic individuals (Fig. 5A).',\n",
       "  'Furthermore, Ara h 2–binding IgG+ B cells from peanut-\\xadallergic individuals were signifi- cantly more likely to be of the IgG1 sub-\\xadisotype (85.7% of IGHG cells; Fig. 5B) compared with DT-\\xadbinding IgG1+ B cells from peanut-\\xadallergic (55.1% of IGHG+ cells, P = 0.0015 by chi-\\xadsquare test) or from non-\\xadallergic individuals (52% of IGHG+ cells, P = 0.006 by chi-\\xadsquare test).',\n",
       "  'These results show that Ara h 2– binding IgG+ B cells from peanut-\\xadallergic individuals have defining characteristics of type 2–marked IgG+ cells from cluster 5: high SHM, predominant IgG1 constant regions, and transcription of FCER2/CD23 as well as IGHE.',\n",
       "  'Among Ara h 2–binding IgG+ cells, we identified a group of conver- gent sequences composed of three IgG1+ B cells (P4EC10, P4AC5, and P6EC2) using ambiguously IGHV3-\\xad30*18 or IGHV3-\\xad30-\\xad5*01 paired with IGHJ6*02, as well as IGKV3-\\xad20*01 and either IGKJ2*01 or IGKJ2*03 (data file S3).',\n",
       "  'In ELISA and biolayer inter- ferometry assays, we found that these antibodies bound with high affinity to Ara h 2',\n",
       "  'We identified three additional se- quence families containing at least two Ara h 2–binding cells each, although these families did not span across multiple individuals.',\n",
       "  'We found that all identified members of the 4C5G1 convergent family bound to the same Ara h 2 epitope (Ara h 2.008; fig. S13B).',\n",
       "  'To summarize, we found that B cell clones that bind the main allergen Ara h 2 with high affinity were predominantly somatically mutated, type 2–marked IgG1 memory cells that transcribed germ- line IGHE, contained convergent BCR sequences, and shared epitope recognition with IgE antibodies.',\n",
       "  'DISCUSSION Using single-\\xadcell transcriptomics and functional analysis, we identified a population of somatically mutated IgG+ memory B cells character- ized by expression of IL-\\xad4– and IL-\\xad13–regulated genes, CD23/ A Isotype IGHG1 IGHG2 IGHG3 IGHG4 B C SHM frequency Ara h 2 DT DT Ara h 2 DT DT Non- allergic Peanut-allergic 0.00 0.25 0.50 0.75 1.00 Proportion of IGHG B cells Ara h 2 DT DT P = 0.0044 P < 0.0001 P = 0.0305 IGHG IGHM 0.0 0.1 0.2 Relative expression ACTB P = 0.0028 FCER2 GLT-IGHE total-IGHE 1.2 1.0 0.8 0.6 0.4 0.2 0.0 1.0 0.8 0.6 0.4 0.2 0.0 1.0 0.8 0.6 0.4 0.2 0.0 DT DT Ara h 2 Non- allergic PsIgE>100 DT DT Ara h 2 Non- allergic PsIgE>100 DT DT Ara h 2 Non- allergic PsIgE>100 P = 0.0150 P = 0.0021 P < 0.0001 P = 0.0001 P < 0.0001 Non- allergic Peanut-allergic Non- allergic Peanut-allergic Fig.',\n",
       "  'We found that the highly sensitized peanut-\\xadallergic children had a much higher frequency of Ara h 2– binding IgG1+ memory B cells than children with low concentrations of PsIgE, again suggesting a relationship between peanut-\\xadspecific IgG1+ memory cells and PsIgE antibodies.',\n",
       "  'Peanut-\\xadallergic children with high concentrations of PsIgE (higher than 100 kUA/liter), low concentrations of PsIgE (lower than 5 kUA/liter), or intermediate concentrations of PsIgE (between 5 to 100 kUA/liter) were recruited among the pediatric population attending the Food Allergy clinic at Mount Sinai Hospitals.',\n",
       "  'Among 58 pediatric patients with peanut allergy, 18 patients had PsIgE higher than 100 kUA/liter, (PsIgE>100), 16 patients had sIgE lower than 5 kUA/liter (PsIgE<5), and 24 patients had sIgE between 8.0 and 76.6 kUA/liter (5 < PsIgE<100).',\n",
       "  'In terms of incidence of other atopic diseases, 89% of peanut-\\xadallergic children with PsIgE>100 and 69% of peanut-\\xadallergic children with PsIgE<5 reported a history of AD or asthma, which was not significantly different between the two groups (P = 0.147; table S1).',\n",
       "  'In our consensus sequences, however, SHM frequencies of paired heavy and light chains were closely correlated (linear regression slope P < 2 × 10−16, R2 = 0.63).'])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hypotheses = []\n",
    "methods = []\n",
    "datasets = []\n",
    "experiments = []\n",
    "analyses = []\n",
    "results = []\n",
    "for ind, sent in enumerate(doc.sents):\n",
    "    sent_text = re.sub(r'\\s+', ' ', sent.text.strip().replace(\"\\n\", \" \"))\n",
    "    if sent_text and any(pattern.search(sent_text) for pattern in compiled_hypothesis_patterns):\n",
    "        hypotheses.append(sent_text)\n",
    "    elif sent_text and any(pattern.search(sent_text) for pattern in compiled_methods_patterns):\n",
    "        methods.append(sent_text)\n",
    "    elif sent_text and any(pattern.search(sent_text) for pattern in compiled_dataset_patterns):\n",
    "        datasets.append(sent_text)\n",
    "    elif sent_text and any(pattern.search(sent_text) for pattern in compiled_experiment_patterns):\n",
    "        experiments.append(sent_text)\n",
    "    elif sent_text and any(pattern.search(sent_text) for pattern in compiled_analysis_patterns):\n",
    "        analyses.append(sent_text)\n",
    "    elif sent_text and any(pattern.search(sent_text) for pattern in compiled_result_patterns):\n",
    "        results.append(sent_text)\n",
    "\n",
    "hypotheses, methods, datasets, experiments, analyses, results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62c1bd36-94f3-4b88-8037-108d0d077c36",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-11T09:53:00.028981Z",
     "start_time": "2025-10-11T09:53:00.016472Z"
    }
   },
   "outputs": [],
   "source": [
    "for ind, sent in enumerate(doc.sents):\n",
    "    sent_text = re.sub(r'\\s+', ' ', sent.text.strip().replace(\"\\n\", \" \"))\n",
    "    print(f\"{ind}: {sent_text}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
